### ISSUE TITLE
[New build - DEFAULT] 20w27c: Flight Model Improvement

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#2616]
The spaceship flight control has been fixed, it is now much more precise and will stop much faster when in the dead zone.

Bugfixes
#2593 #2592 Piston orientation.
#2580 Landing pad on pad with a relative low weight dont finish.
#2558 Repair at shipyard dont reset internal brick hp state.
#2604 FPS info not displaying with UI Post Effect On.
New Weapon start message in multiplayer.
Weapon unspawn while on Item Stand cause issues.
#2605 Piston can have negative values.
#2606 #2585 Shipyard undocking causes issue.
#2540 Can link several sequencer to a mechanism.
#2611 Piston incorrect state on reload.

**Comment by tsunamayo** [comment]
Build is up! Tell me how you feel about the new control. Does it deserve a new control scheme? 
Full stroy: I tried the one from Avorion / War Thunder, it was an absolute garbage because of the physics part of the code. Then I fixed the physics and made it worked, but it also fixed the standard classic control... So lets try this one first.
Cheers

**Comment by Brushes2** [comment]
Feels great! I did notice though, when performing more complex maneuvers, the ships tend to become sluggish. Try going full speed and pulling up while rolling. You'll notice the rates change suddenly and im not sure if that's a function of the flight computer or ship designs.

You mentioned War Thunder's setup; any plans to add adjustable dead zones or input curves?

**Comment by TripleHelixxx** [comment]
Only one sequencer two a mechanism? RIP Multi-position chairs. 

The flight control did feel a little better though, so I guess there’s that.

**Comment by ProPeach** [comment]
Happy to test!
Flying definitely feels tighter - when you move your mouse back to the centre after manoeuvring, there's less danger of overshooting and the ship overcompensating - but there's still a long way to go with regards to the deazone as Brushes mentions, as well as the ship following the mouse in a logical sense

I've used this comparison before, but I hope I can be a bit clearer this time
Here is what happens when you move your mouse in Elite Dangerous - https://i.imgur.com/HbRjUi7.mp4
The moment you move your mouse away from the centre, the ship starts following it. Very slowing at first, but with a nice linear acceleration in that exact direction. The deadzone is tiny, almost unnoticeable. If you move your mouse diagonally, the ship follows it diagonally. You can see how smooth and natural this feels, I feel like I have complete control over my ship. Honestly Elite feels like the best model to be taking notes from, it's widely regarded to have absolutely nailed the *feeling* of flying a ship, which is very hard to capture using a mouse and keyboard.


Here is the same test in Starship EVO with the current flight model - https://i.imgur.com/HJVWoX4.mp4
In comparison it feels very rigid and forced. I think the main culprits are the huge, square deadzone as well as the tendency for a ship to pitch *or* yaw with small movements, but never both. This makes it feel like your mouse movements are adapted into a joystick, and then fed into the game which is not comfortable. Ideally the ship should follow the direction of the mouse smoothly like Elite with no unexpected direction changes, so a diagonal mouse movement means a diagonal ship movement. You can really tell the difference at the end of both clips where I move the mouse in a circular motion; Elite follows the mouse smoothly, while in the current model the mouse doesn't even leave the deadzone.
Here's the ship I used in this clip if you'd like to replicate it exactly - [A-Wing - ProPeach.zip](https://github.com/tsunamayo/Starship-EVO/files/4869931/A-Wing.-.ProPeach.zip)

Getting this model right will probably be tough, but we'll help you out! I do think that relying on the community a bit more in this instance would net the best results as this will likely require quite a lot of time and effort.

**Comment by T00Code34215** [comment]
Anyone know how to un-dock from ships/stations

**Comment by Kaiser-Indrasil** [comment]
Feels much better than the previous one - I can finally do a fast reverse manoeuvre without overshooting!

I agree with Peach, Elite's model vastly superior over everything else we have on the market. Starship Evo's only 8-directional pitch and yaw steering in low attack angles, and an overly huge deadzone are what I believe to be the main culprits of why it feels sluggish and unwieldy. It feels like I'm moving with a digital-input 8-directional pad instead of an analogue-input mouse.

What it feels like in the game:
![flight model - current](https://user-images.githubusercontent.com/63756057/86491048-6f007600-bd61-11ea-887e-598c062222b9.png)
Very big deadzone in the centre, middle ring of only 8-directional yaw/pitch adjustment, and outer ring where it starts to feel more gradual, but not quite - you can still feel something like mouse angle snapping near the X and Y axes, which feels bad:
![image](https://user-images.githubusercontent.com/63756057/86491243-285f4b80-bd62-11ea-9130-67678e8ebc88.png)

What we want and what Elite's flight model feels like:
![flight model - desired](https://user-images.githubusercontent.com/63756057/86491269-390fc180-bd62-11ea-9ec0-06d5b73b07e6.png)
Very small deadzone and full analogue input. No approximation, no angle snapping. The craft pitches and yaws at exact angle you want, even if it's 1 degree off an axis.

@T00Code34215 This is certainly not the place to ask about that. We will gladly answer any question we know the answer for on Discord in the #support channel.

**Comment by StarrKiss** [comment]
I agree with @Kaiser-Indrasil , beautiful visualization. Elite control scheme would be wunderbar.

**Comment by tsunamayo** [comment]
Okay Inow added a much tighter dead-zone (still needs one otherwise you can fly straight). Thanks for the feedback and the video/pictures, very informative!
For the "angle snapping", I have no idea on what that could be, I dont really feel this on my side and there is nothing like in the code... Does other people felt this also?

**Comment by tsunamayo** [comment]
okay so I think that "angle snapping" is just the square dead zone, when you move close to it you only get a single axis moving up until you are close a corner. I will implement a round one, the one from elite seems round also (not elliptical)? Cheers

**Comment by Kaiser-Indrasil** [comment]
Yes, the angle snapping is exactly that. I'm keeping my fingers crossed for the improvements you're going to introduce!

**Comment by ProPeach** [comment]
Sounds brilliant! Do you still plan on keeping manoeuvre and boost mode separate? How about a throttle implementation? Glad we could help : )

**Comment by tsunamayo** [comment]
yeah we will see for throttle, I want a solid control scheme first, I think I will have to rethink all of it, weapon firing included...

**Comment by tsunamayo** [comment]
Things I want to re-examine:
- Selecting target
- Selecting action on target
- Firing weapon
- selecting weapon or weapon group
- free look key

**Comment by Kaiser-Indrasil** [comment]
My thoughts on each of these points:
- target selecting could work much like it works now, with the middle mouse button - it's very easy to reach and does not get in the way of steering your ship.
- what actions would you like to implement?
- weapon firing with left and right mouse button feels good as it is. What concerns you in that regard?
- weapon groups could be placed on the yoke's radial menu, which would have to be implemented. The player would click the Mouse 4 button (one of the thumb buttons. I'm sure almost everyone has them at this point. Aside from Mac users, but who plays on a Mac...) to open up the Weapons Group radial menu, hover over the desired group, and click either the Left Mouse button or the Right Mouse button, depending on which button would they want to use at the time. Something like GTA V or Doom but you get up to 2 weapon groups to shoot at the given time. Let the user configure each individual groups with logic gates first and let them link the groups to the yoke.
- Free look key could be the Mouse 5 button - so we could maintain straight line fire and manoeuvreing with our WASD keys and also free look at the same time. Short click - toggle the free look mode. Long click (more than 1 second) - maintain free look only until the button is pressed. Both clicks would save the new camera position. Double click - reset camera position to default. Not sure if that's the best way but sounds convenient enough. I like to have a different camera angle in some games sometimes while not having to hold the button constantly.

**Comment by tsunamayo** [comment]
okay deadzone fixed - it is now circular.
I think I will open a discussion for the control. I think having a Mouse 5 by default seems like a stretch, Middle Mouse seems to be the maximum but I am not sure what other game are doing.

**Comment by Kaiser-Indrasil** [comment]
That's great to hear! Can't wait to test it in the next patch!
As for the other games, I played flight sims like Star Citizen, Elite: Dangerous, Freespace 2, and other space games like Rebel Galaxy Outlaw, Avorion, Space Engineers as prime examples. They all have different button assigned to targeting different kinds of ships (enemy; friendly; nearest enemy; in front of you; your target's attacker; your target's target etc.) Whilst all these ways of targeting are incredibly useful, they take a lot of time and effort to learn and master and that's not really our aim, I believe.

Simple target-the-thing-in-front-of-you system will do for now, and I think the middle mouse button will be a good fit. It's out-of-the-way and easy to reach.

**Comment by StarrKiss** [comment]
I would love the addition of a throttle
